Store WordPress "build" directory as an artifact to enable live previews with Playground

adamziel opened the pull request:
Creates a new GitHub workflow to expose WordPress build as a GitHub artifact. The goal is to enable previewing Pull Requests in WordPress Playground.

Related Playground PR https://github.com/WordPress/wordpress-playground/pull/700

Trac Ticket: https://core.trac.wordpress.org/ticket/59416

cc @gziolo @youknowriad @peterwilsoncc @desrosj

adamziel committed:
Store WordPress "build" directory as an artifact to enable live previews with Playground

adamziel committed:
Upload artifacts in e2e workflow, not test-npm workflow

adamziel committed:
No, use test-npm.ml after all :D

adamziel committed:
Upload the build directory

adamziel committed:
Adjust paths

jeffpaul commented:
Note that this is a foundational item for https://core.trac.wordpress.org/ticket/59416 and will hopefully make it dramatically easier for folks to help test PRs.  Thanks for the work on this @adamziel.  This is so great, I LOVE IT!!!

desrosj commented on .github/workflows/end-to-end-tests.yml:
```diff
@@ -134,6 +134,14 @@ jobs:
+        uses: actions/upload-artifact@0b7f8abb1508181956e8e162db84b466c27e18ce # v3.1.2
+        if: always()
+        with:
+            name: wordpress-build
```
Can we have more specific artifact names? If this is for each PR, we should probably append the PR number. Or we could be more specific and use commit SHA.

desrosj commented on .github/workflows/end-to-end-tests.yml:
```diff
@@ -134,6 +134,14 @@ jobs:
+            name: wordpress-build
+            path: build
+            if-no-files-found: ignore
+
```
I think we should move this after the `git diff` check. If that returns an error, we shouldn't create an artifact as there are changes that must be committed first.

desrosj commented on .github/workflows/end-to-end-tests.yml:
```diff
@@ -134,6 +134,14 @@ jobs:
+        with:
+            name: wordpress-build
+            path: build
+            if-no-files-found: ignore
```
We should probably use `error` here. If the artifact fails to upload, that's a problem.

Also, failed workflows automatically restart once. So this would force a failure and retry to be sure there's not something else going on.

desrosj reviewed (commented):
I'm a little confused as to why we're uploading the artifacts 3 times. Could you clarify?

Also, I'm wondering if there is a benefit to create a new workflow that's specifically meant to just build the package and upload it as an artifact. That would open the door to perform any other additional steps we may need down the road as well.

adamziel committed:
Use a dedicated workflow to build WordPress

adamziel commented:
> I'm a little confused as to why we're uploading the artifacts 3 times. Could you clarify?

@desrosj that's what I meant by "it's overeager and tries to upload multiple times" in the description. No purpose in particular, I just ran out of time to fix it on Friday. I just corrected that so it only uploads the artifact once. I also applied the rest of your feedback, thank you!

Right now I'm trying to figure out why GitHub is not running the workflow after I moved it to a separate `build.yml` file.

adamziel committed:
Move the build workflow to test-npm.yml – will it work now?

adamziel committed:
Adjust the workflow name

adamziel committed:
Move the workflow back to a separate file

desrosj commented:
> Right now I'm trying to figure out why GitHub is not running the workflow after I moved it to a separate `build.yml` file.

Hmm, looking at the commit you made above, I'm not sure either. The `paths` under the `pull_request` event are probably unnecessary. We want each change to trigger a build. With those `paths` defined, we may not get an artifact every time.

Another thing I've been thinking about. Will these artifacts need to be available more widely? I think the current intent is to share a link within the context of a PR. But if we'd need to use the artifact anywhere else, I'm not sure if it's easy to retrieve the artifact. it may be worth looking at publishing these as a [package on GitHub](https://github.com/orgs/WordPress/packages), or something similar. This may be a problem for a different day, though.

adamziel commented:
> But if we'd need to use the artifact anywhere else, I'm not sure if it's easy to retrieve the artifact.

It's [not easy](https://github.com/WordPress/wordpress-playground/blob/trunk/packages/playground/website/public/plugin-proxy.php#L41) and I'd love to make it easy at one point. At the same time, artifacts solve a lot of problems, like cleaning up old files. I'd say let's move forward with them for now and let's explore more accessible uploads separately.

adamziel commented:
Also, I found the problem with a separate workflow file. I forgot about the required `runs-on` option. I fixed it and the build workflow now works as expected. [Example run](https://github.com/WordPress/wordpress-develop/actions/runs/6534518101?pr=5481).

adamziel committed:
Zip the WordPress build to hopefully make the upload faster

adamziel commented:
And I just added a `zip -r build.zip build` line that sped up the upload step from 10 minutes to 19s. Woah!

Does this look good, or do you see any blockers @desrosj?

desrosj commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,84 @@
+permissions: {}
+
+jobs:
+  build-wordpress:
```
Can we add a documentation block here that has a description for the job and a list of steps that it will take? See some other workflows for an example.

desrosj commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,84 @@
+        uses: actions/upload-artifact@0b7f8abb1508181956e8e162db84b466c27e18ce # v3.1.2
+        if: always()
+        with:
+            name: wordpress-build-${{ github.sha }}
```
Thinking about this more, I think it would make sense to include the PR number in the artifact name because SHAs are not human readable. In the future, it would be nice if all artifacts for a given PR share the same name and the most recent one is always returned.

```suggestion
            name: wordpress-build-${{ github.event_name == 'pull_request' && github.event.number && '-' || github.sha }}
```

desrosj commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,84 @@
+  push:
+    branches:
+      - trunk
+      - '3.[7-9]'
```
Thinking about this more as well. Do we want artifacts to be present for every commit and every branch/tag going forward? If so, then this is fine. But if we only need them for pull requests right now, we should remove the `push` event here, and probably the branch filters below for `pull_request`.

desrosj commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,84 @@
+      - trunk
+      - '3.[7-9]'
+      - '[4-9].[0-9]'
+    paths:
```
If we need an artifact for every pull request, these path filters should be removed. otherwise there will be PRs where this never runs.

desrosj reviewed (changes requested):
Sorry, a few more thinks that I noticed!

desrosj commented:
Added a few small requests. Also, thinking about this more, if the intent is for this workflow to run on any branch, we will need to backport this and should wait until after WP 6.4 is branched to commit to avoid having to backport frequent changes to the 6.3 branch as we refine.

I think it's fair to say we should only run this for PRs to and commits to `trunk` for now. Added a comment inline about that.

desrosj commented:
> And I just added a `zip -r build.zip build` line that sped up the upload step from 10 minutes to 19s. Woah!

One thing to keep in mind here is that this results in the zip file that is uploaded as an artifact being zipped again. It's not immediately apparent, but it may require an additional step when using the artifact.

adamziel committed:
Document the purpose of the workflow

adamziel committed:
Update .github/workflows/build.yml

Co-authored-by: Jonathan Desrosiers <desrosj@users.noreply.github.com>

adamziel committed:
Remove "on push:" and "paths:" filters to create these artifacts for all PRs

adamziel committed:
Remove branch filter for pull_request

adamziel commented:
@desrosj good feedback, thank you! I just addressed all your notes.

> I think it's fair to say we should only run this for PRs to and commits to trunk for now. Added a comment inline about that.

Yes, I only want to run this for PRs so we're good.

> One thing to keep in mind here is that this results in the zip file that is uploaded as an artifact being zipped again.

This is fine, Playground can handle that – the Gutenberg repo does the same thing.

adamziel committed:
Adjust ternary to use PR number

ockham commented:
>> And I just added a `zip -r build.zip build` line that sped up the upload step from 10 minutes to 19s. Woah!

That's due to a somewhat [familiar issue](https://github.com/actions/upload-artifact/issues/199) with the `upload-artifact` action: Uploading multiple files takes significantly more time than uploading a single file 😕 

>One thing to keep in mind here is that this results in the zip file that is uploaded as an artifact being zipped again. It's not immediately apparent, but it may require an additional step when using the artifact.

FWIW, the same tradeoff applies to Gutenberg (where it has confused users that uploaded the "double-zipped" `gutenberg-plugin.zip` to their WP, expecting it to work like a "regular" plugin zip file); there's also [an issue in the `upload-artifact` repo](https://github.com/actions/upload-artifact/issues/39) about the double-zip. We even attempted removing that extraneous zipping at some point (https://github.com/WordPress/gutenberg/pull/32780) but concluded that it wasn't worth the negative performance impact.

ockham commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,67 @@
+        run: npm run build
+
+      - name: Zip WordPress build
+        run: zip -r build.zip build/*
```
Should we maybe name the output file `wordpress.zip` instead? After all, it will be a fully functional WP install, based on the current branch, no?

(Really looking forward to this; this will come in very handy for testing! 🎉 )

adamziel replied:
Just renamed it @ockham!

adamziel committed:
Rename build.zip to wordpress.zip

ockham commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,67 @@
+        run: npm run build
+
+      - name: Zip WordPress build
+        run: zip -r wordpress.zip build/*
```
I just downloaded the artifact produced by this PR and unzipped it (twice). The result still contains the `build/` folder (rather than having all files at the root).

Can we change this to only include the files in `build/`, but not the folder itself? Maybe the following would do the trick?

```suggestion
        run: zip -r wordpress.zip build/
```

ockham replied:
> Maybe the following would do the trick?
> [...]

Ah no, that doesn't work. But 

```suggestion
        run: cd build && zip -r wordpress.zip .
```

should probably, or alternatively

```suggestion
        working-directory: build
        run: zip -r wordpress.zip .
```

(We'll also need to update the `path:` for the `upload-artifact` step, then.)

adamziel committed:
Update .github/workflows/build.yml

Co-authored-by: Bernie Reiter <96308+ockham@users.noreply.github.com>

ockham commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,67 @@
+        if: always()
+        with:
+            name: wordpress-build-${{ github.event_name == 'pull_request' && github.event.number || github.sha }}
+            path: wordpress.zip
```
```suggestion
            path: build/wordpress.zip
```

adamziel committed:
Update .github/workflows/build.yml

Co-authored-by: Bernie Reiter <96308+ockham@users.noreply.github.com>

ockham reviewed (approved):
LGTM!

Shall we wait for 6.4 to be branched from `trunk` before landing this? 🤔 (That should happen shortly after RC1, which is scheduled for later tonight.)

adamziel commented:
> Shall we wait for 6.4 to be branched from trunk before landing this?

This change matters for trunk Pull Requests and even if the workflow file changes, I don't think we'll need to backport that. I'll go ahead and merge. Feel free to revert if I'm wrong here.

adamziel commented:
https://core.trac.wordpress.org/changeset/56958
